### PR TITLE
Fix: Prompt logprobs misaligned: left shift needed; BOS logprob must be NaN

### LINKED ISSUE #2103: Prompt logprobs misaligned: left shift needed; BOS logprob must be NaN
### Bug report

Observed **very large negative logprobs across all prompt positions**. 
Root cause: in maxengine.py prefill() line 478: 
`prompt_logp = inference_utils.log_prob_of_chosen_token(flat_logits, input_tokens)`
logprobs were computed as **log P(token[t] | tokens[0..t])** (same-position logits), rather than **log P(token[t] | tokens[0..t-1])**. This off-by-one alignment makes real tokens look extremely unlikely.

### Details
- Decoder logits at step *i* predict token *i+1*. To score token *t*, we must use **logits[t-1]**, i.e., shift by one.
- The first prompt position (BOS) has no preceding context, so its logprob must be **null/NaN** (OpenAI semantics).
- Padding/final positions should not be scored; the final prefill logit predicts the **first completion token**, not a prompt token.

### Expected behavior
- Return per-token prompt logprobs: `log P(token[t] | tokens[:t])` for `t = 1..true_length-1`.
- `logprobs[0] = NaN` (BOS).
- Positions `>= true_length` are ignored (NaN).

### Fix outline
- Left-shift targets: score tokens `1..S-1` using logits `0..S-2`.
- Mask `t=0` (BOS) and any `t >= true_length` to NaN.
- Keep using the final prefill logit (`t = true_length-1`) **only** to score the **first generated token**.

### Input Prompt: Hello, who are you?  -->` ['', 'Hello', ',', 'who', 'are', 'you', '?']`
Before: Prompt logprobs
image

### Logs/Output

_No response_

### Environment Information

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by babyplutokurt**
Fixes #2103 

Observed **very large negative logprobs across all prompt positions**. 
Root cause: in maxengine.py prefill() line 478: 
`prompt_logp = inference_utils.log_prob_of_chosen_token(flat_logits, input_tokens)`
logprobs were computed as **log P(token[t] | tokens[0..t])** (same-position logits), rather than **log P(token[t] | tokens[0..t-1])**. This off-by-one alignment makes real tokens look extremely unlikely.

### Details
- Decoder logits at step *i* predict token *i+1*. To score token *t*, we must use **logits[t-1]**, i.e., shift by one.
- The first prompt position (BOS) has no preceding context, so its logprob must be **null/NaN** (OpenAI semantics).
- Padding/final positions should not be scored; the final prefill logit predicts the **first completion token**, not a prompt token.

### Expected behavior
- Return per-token prompt logprobs: `log P(token[t] | tokens[:t])` for `t = 1..true_length-1`.
- `logprobs[0] = NaN` (BOS).
- Positions `>= true_length` are ignored (NaN).

### Fix outline
- Left-shift targets: score tokens `1..S-1` using logits `0..S-2`.
- Mask `t=0` (BOS) and any `t >= true_length` to NaN.
- Keep using the final prefill logit (`t = true_length-1`) **only** to score the **first generated token**.

## Input Prompt: Hello, who are you?  -->` ['', 'Hello', ',', 'who', 'are', 'you', '?']`
Before: Prompt logprobs
image

After Fix: 
Screenshot 2025-08-07 6 53 40 PM

### CODE REVIEW

File: MaxText/maxengine.py [hunk 1]
```diff
@@ -475,7 +475,7 @@ def _prefill_jit(
           page_state=page_state,
       )
     if return_prompt_logp:
-      prompt_logp = inference_utils.log_prob_of_chosen_token(flat_logits, input_tokens)
+      prompt_logp = inference_utils.prompt_logprobs_from_prefill(flat_logits, input_tokens, true_length)
```
Comment (RissyRan) [inline MaxText/maxengine.py:478]: I see log_prob_of_chosen_token is used 2 places in maxengine.py ([1](https://github.com/AI-Hypercomputer/maxtext/blob/bd33a7c9003702a457a04ebbe596d93de6fa372c/MaxText/maxengine.py#L686), [2](https://github.com/AI-Hypercomputer/maxtext/blob/bd33a7c9003702a457a04ebbe596d93de6fa372c/MaxText/maxengine.py#L503)). Shall we update both?

**Comment by patemotter** [comment]
There are other calls in `maxengine.py` to the original `log_prob_of_chosen_token` code. Do any of those need to be migrated as well? Specifically `prefill_concat()`.

**Comment by babyplutokurt** [comment]
> There are other calls in `maxengine.py` to the original `log_prob_of_chosen_token` code. Do any of those need to be migrated as well? Specifically `prefill_concat()`.

Modified logic in prefill_concat(), we need to shift and calculate the actual logprobs for each token, then mask invalid position with Nan

**Comment by patemotter** [comment]
Overall LGTM. The only thing I would add is relevant unit tests for these new functions into `maxtext_utils_test.py`

**Comment by babyplutokurt** [comment]
> Overall LGTM. The only thing I would add is relevant unit tests for these new functions into `maxtext_utils_test.py`

Added

File: MaxText/maxengine.py (hunk 1 above)
Comment (RissyRan) [reply MaxText/maxengine.py:478]: Oh... I haven't pushed this out. But please double check if need.

**Review by RissyRan — APPROVED** [review]
LGTM! Please address unresolved comments before the merge. Thanks!
